crates/sui-move-build/src/lib.rs

```diff
@@ -184,10 +188,22 @@ impl BuildConfig {
 
     /// Given a `path` and a `build_config`, build the package in that path, including its dependencies.
     /// If we are building the Sui framework, we skip the check that the addresses should be 0
-    pub fn build(self, path: PathBuf) -> SuiResult<CompiledPackage> {
+    pub fn build(mut self, path: PathBuf) -> SuiResult<CompiledPackage> {
```

Comment:
This raises a good points! High-level, I'd assumed that we want a top-level (root) package to unconditionally use flags in the `Move.lock` file, but I see now how that can conflict with unsetting/implicit flags using this PRs "config merge" logic. 

> E.g. if I did a move build before, and set certain configs (for flavor and edition), and then did a subsequent move build without those flags, then the previously used flavor and edition would be remembered and used. Is that right?

Right, so, that's not the intention. (And sorry that this PR is unclear about the other half of the picture where we decide what / when to write to `Move.lock`). On that, the original thinking is still that when a `move build` succeeds (with whatever flags / compiler version) then the `Move.lock` is updated to capture that config.

To your point, we should only need the `Move.lock` info in specific cases (the circumstances you mention). I agree its wrong to use (possibly previous) `Move.lock` values when a user subsequently invokes command line that does not explicitly specify (or unsets) flags.

I'm going to revert this PR to draft and rework it for the specific cases. I'll first add the bit that writes the `Move.lock` content, since I think there's a clear path there, and come back to this.

![](https://media.giphy.com/media/G7iGNzr3VBING/giphy.gif)